### PR TITLE
Bug fixes for query builder filtering with ISIN filters over sets with duplicate str representations

### PR DESCRIPTION
**Pull request by poodlewars**
## Bug 1

Fix for OR filters with isin with value sets with clashing str representations:

```
q = q[(q["a"].isin(vals1)) | (q["a"].isin(vals2))]
```

str(np.ndarray) truncates like [0, 1, 2, ... 99997, 99998, 99999] so it is easy for ValueSet str representations to clash. We have special handling for this to suffix -vX in the ValueSet leaf node, but there was no such handling for the ExpressionNode,

```
ISIN([VALUE_SET_1])
```

This meant that filters,

```
q["a"].isin(V1) | q["a"].isin(V2)
```

would incorrectly create a single node key, `ISIN [1, 2 ... 999]` and one of the two filters would effectively be dropped at evaluation time, since we used these strings to describe the edges in the AST.

## Bug 2

Drop release notes for this part - not user facing as column stats is feature flagged. Fixes expressions like this with column stats in the case where the two value sets have the same `str` representation:

```
    q = QueryBuilder()
    q = q[q["col_1"].isin(value_set_one)]
    q = q[q["col_1"].isnotin(value_set_two)]
```

Column stats AND-s filters together in to one in `and_filter_expression_contexts`., which merged the trees in two ExpressionContext objects together in to a graph, sharing leaf nodes where they matched on names.

These names in the original filter AST come from this in `processing.py`:

```
def to_string(leaf):
    if isinstance(leaf, (np.ndarray, list)):
        # Truncate value set keys to first 100 characters
        key = str(leaf)[:100]
    elif isinstance(leaf, _RegexGeneric):
        key = "Regex({})".format(leaf.text())
    else:
        if isinstance(leaf, str):
            key = "Str({})".format(leaf)
        elif isinstance(leaf, bool):
            key = "Bool({})".format(leaf)
        else:
            key = "Num({})".format(leaf)
    return key
```

so different ValueSets can clash, leading to an incorrect merged tree . Note `str(np.ndarray)` also truncates like `[0, 1, 2, ... 99997, 99998, 99999]`, so the "first 100 chars" is not the only problem.

## Fix

We now form the query AST as a graph in the frontend, rather than relying on string references to encode edges. This gets rid of the need to make the two fixes below.

We keep the concept of memoizing results, ie making sure we only evaluate subtrees once in cases like:

```
  # only want to calculate the isin once
  big = list(range(100_000))
  q = q[q["id"].isin(big) & (q["t"] > 0) | q["id"].isin(big)]

  # only want to calculate the spread once
  spread = q["bid"] - q["ask"]
  q = q[(spread > 0) & (spread < limit)]
```

We do this by looking up expressions that we are going to evaluate in a map of string label name to expression node and result. To guard against conflicts, we do a deep comparison against the node that created the result.

### CODE REVIEW

File: python/tests/unit/arcticdb/test_column_stats_isin.py [hunk 1]
```diff
@@ -504,3 +504,136 @@ def test_column_stats_isin_int64_min_not_treated_as_nat(
...
+    qs.reset_stats()
+    result = lib.read(sym, query_builder=q).data
+    table_data_reads = get_table_data_read_count()
+
+    assert_frame_equal(df0, result)
+    expected_reads = 1
+    assert table_data_reads == expected_reads, f"Expected {expected_reads} TABLE_DATA read(s), got {table_data_reads}"
+
+
+def test_column_stats_isin_and_isnotin_colliding_value_sets_nonreg(
```
Comment (poodlewars) [inline python/tests/unit/arcticdb/test_column_stats_isin.py:600]: This is the only test that was failing before the change. I think the ones above are worth adding anyway.

File: cpp/arcticdb/processing/expression_context.hpp [hunk 2]
```diff
@@ -9,71 +9,21 @@
 #pragma once
 
 #include <memory>
-#include <string>
-#include <unordered_map>
 
 #include <arcticdb/processing/expression_node.hpp>
-#include <arcticdb/pipeline/value.hpp>
-#include <arcticdb/pipeline/value_set.hpp>
+#include <arcticdb/util/preconditions.hpp>
```
Comment (alexowens90) [inline cpp/arcticdb/processing/expression_context.hpp]: ?

File: cpp/arcticdb/processing/expression_node.hpp [hunk 3]
```diff
@@ -85,31 +69,56 @@ struct BitSetTag {};
...
-    OperationType operation_type_;
+    using Leaf = std::variant<
+            ColumnName, std::shared_ptr<Value>, std::shared_ptr<ValueSet>, std::shared_ptr<util::RegexGeneric>>;
+
+    struct Operation {
+        OperationType operation_type_;
+        std::shared_ptr<ExpressionNode> condition_;
+        std::shared_ptr<ExpressionNode> left_;
+        std::shared_ptr<ExpressionNode> right_;
+    };
```
Comment (alexowens90) [inline cpp/arcticdb/processing/expression_node.hpp:89]: nit: I think it would look slightly cleaner if the member ordering was `operation_type_, left_, right_, condition_`, with a ctor with defaulted `nullptr` values for `right_` and `condition_`

File: cpp/arcticdb/processing/expression_node.cpp [hunk 4]
```diff
@@ -75,43 +76,186 @@ ColumnWithStrings::ColumnWithStrings(
...
+            [&seg, this](const Operation& op) -> VariantData {
+                if (auto it = seg.computed_data_.find(label_); it != seg.computed_data_.end()) {
+                    const auto& [other, cached] = it->second;
+                    if (other == this || nodes_equal(*this, *other)) {
+                        log::version().debug("Memoization hit for {}", label_);
+                        return cached;
+                    }
+                    log::version().debug("Memoization label collision for {}", label_);
+                }
+                log::version().debug("Memoization miss, computing {}", label_);
```
Comment (alexowens90) [inline cpp/arcticdb/processing/expression_node.cpp]: This will be very noisy, I'd either remove or compile out of debug builds

File: cpp/arcticdb/processing/expression_node.cpp [hunk 5]
```diff
@@ -75,43 +76,186 @@ ColumnWithStrings::ColumnWithStrings(
...
+                        );
+                    } else if (is_binary_operation(op.operation_type_)) {
+                        return dispatch_binary(op.left_->compute(seg), op.right_->compute(seg), op.operation_type_);
+                    } else {
+                        return dispatch_unary(op.left_->compute(seg), op.operation_type_);
+                    }
+                }();
+                // On a label collision the slot is already held by a structurally-different node; try_emplace is a
+                // no-op there, so that node keeps the slot and this result simply isn't memoized.
+                seg.computed_data_.try_emplace(label_, this, result);
```
Comment (alexowens90) [inline cpp/arcticdb/processing/expression_node.cpp]: I think it _is_ worth logging if this emplace fails, as you could conceivably see unexpected bimodal performance for similar looking queries in this case

File: cpp/arcticdb/processing/expression_node.cpp [hunk 6]
```diff
@@ -75,43 +76,186 @@ ColumnWithStrings::ColumnWithStrings(
...
+        return fmt::format("{}({})", op.operation_type_, op.left_->label_);
+    }
+}
+
+bool value_sets_equal(ValueSet& a, ValueSet& b) {
+    if (a.empty() != b.empty()) {
+        return false;
+    }
+    if (a.base_type() != b.base_type()) {
+        // This might cause some false negatives but the simplicity is worth it
```
Comment (alexowens90) [inline cpp/arcticdb/processing/expression_node.cpp:115]: Agreed

File: cpp/arcticdb/processing/expression_node.cpp [hunk 7]
```diff
@@ -75,43 +76,186 @@ ColumnWithStrings::ColumnWithStrings(
...
+        const auto& leaf_b = std::get<ExpressionNode::Leaf>(b.kind_);
+        if (leaf_a.index() != leaf_b.index()) {
+            return false;
+        }
+        // Now we know that leaf_a and leaf_b hold the same variant type member
+        return util::variant_match(
+                leaf_a,
+                [&leaf_b](const ColumnName& a_col) { return a_col.value == std::get<ColumnName>(leaf_b).value; },
+                [&leaf_b](const std::shared_ptr<Value>& a_val) {
+                    return *a_val == *std::get<std::shared_ptr<Value>>(leaf_b);
```
Comment (alexowens90) [inline cpp/arcticdb/processing/expression_node.cpp:156]: SImilarly to the false negatives comment above, this wouldn't consider 2 semantically equal but representatively different values as equal (e.g. `np.uint8(0)` and `np.uint16(0)`, but agree it isn't worth worrying about

File: cpp/arcticdb/processing/expression_node.cpp [hunk 8]
```diff
@@ -75,43 +76,186 @@ ColumnWithStrings::ColumnWithStrings(
...
+                [&leaf_b](const std::shared_ptr<ValueSet>& a_set) {
+                    return value_sets_equal(*a_set, *std::get<std::shared_ptr<ValueSet>>(leaf_b));
+                },
+                [&leaf_b](const std::shared_ptr<util::RegexGeneric>& a_regex) {
+                    return a_regex->text() == std::get<std::shared_ptr<util::RegexGeneric>>(leaf_b)->text();
+                }
+        );
+    }
+    const auto& oa = std::get<ExpressionNode::Operation>(a.kind_);
+    const auto& ob = std::get<ExpressionNode::Operation>(b.kind_);
```
Comment (alexowens90) [inline cpp/arcticdb/processing/expression_node.cpp]: nit: `operation_a` and `operation_b`

File: cpp/arcticdb/processing/expression_node.cpp [hunk 9]
```diff
@@ -75,43 +76,186 @@ ColumnWithStrings::ColumnWithStrings(
...
+            [](const std::shared_ptr<Value>& value) {
+                return details::visit_type(value->data_type(), [&value](auto tag) {
+                    using info = ScalarTypeInfo<decltype(tag)>;
+                    return fmt::format(
+                            "Val({}:{})", value->data_type(), value->template to_string<typename info::RawType>()
+                    );
+                });
+            },
+            [](const std::shared_ptr<ValueSet>& value_set) {
+                return fmt::format("ValueSet({},n={})", value_set->base_type().data_type(), value_set->size());
```
Comment (alexowens90) [inline cpp/arcticdb/processing/expression_node.cpp:94]: Could include a few sample values, although the lack of stable ordering could make it confusing

File: cpp/arcticdb/processing/expression_node.cpp [hunk 10]
```diff
@@ -75,43 +76,186 @@ ColumnWithStrings::ColumnWithStrings(
...
+            },
+            [](const std::shared_ptr<util::RegexGeneric>& regex) { return fmt::format("Regex({})", regex->text()); }
+    );
+}
+
+std::string label_for(const ExpressionNode::Operation& op) {
```
Comment (alexowens90) [inline cpp/arcticdb/processing/expression_node.cpp:100]: Guessing this works because by construction child nodes are always created before parent nodes?

File: cpp/arcticdb/processing/processing_unit.cpp [hunk 11]
```diff
@@ -48,56 +48,29 @@ void ProcessingUnit::truncate(size_t start_row, size_t end_row) {
...
-                    auto data = expr->compute(*this);
-                    computed_data_.try_emplace(expression_name.value, data);
-                    return data;
-                }
-            },
-            [&](const std::monostate&) -> VariantData {
-                util::raise_rte("ProcessingUnit::get called with monostate VariantNode");
-            }
-    );
+    if (expression_context_ && !expression_context_->dynamic_schema_) {
```
Comment (alexowens90) [inline cpp/arcticdb/processing/processing_unit.cpp:66]: Is this the only place the `dynamic_schema_` member of `ExpressionContext` is used? If so, clauses know whether dynamic schema is enabled or not via the `set_processing_config` method, so we could scrap `ExpressionContext` entirely

File: cpp/arcticdb/processing/processing_unit.hpp [hunk 12]
```diff
@@ -47,7 +45,12 @@ struct ProcessingUnit {
     std::optional<std::vector<uint64_t>> entity_fetch_count_;
 
     std::shared_ptr<ExpressionContext> expression_context_;
-    std::unordered_map<std::string, VariantData> computed_data_;
+
+    /*
+     * Memoization of results after applying an ExpressionNode to this processing unit. Keyed on the node's label_ with
+     * a deep comparison on hits. One slot per label, colliding writes are dropped.
+     */
+    std::unordered_map<std::string, std::pair<const ExpressionNode*, VariantData>> computed_data_;
```
Comment (alexowens90) [inline cpp/arcticdb/processing/processing_unit.hpp:56]: Maybe worth a comment that the shared pointer owning the `ExpressionNode` is guaranteed to outlive this map, and so the raw pointer is safe

File: cpp/arcticdb/processing/query_planner.cpp [hunk 13]
```diff
@@ -33,34 +33,21 @@ ExpressionContext and_filter_expression_contexts(
...
     util::check(!expression_contexts.empty(), "Expression context cannot be empty");
-    std::optional<ExpressionName> overall_root_name;
-
-    ExpressionContext res;
-    for (auto&& [idx, expression_context] : folly::enumerate(expression_contexts)) {
+    std::shared_ptr<ExpressionNode> root;
+    for (const auto& expression_context : expression_contexts) {
         util::check(
-                std::holds_alternative<ExpressionName>(expression_context->root_node_name_),
+                expression_context->root_ && expression_context->root_->is_operation(),
```
Comment (alexowens90) [inline cpp/arcticdb/processing/query_planner.cpp:40]: I think the second condition is wrong, you can filter directly on a bool column for example

File: cpp/arcticdb/processing/query_planner.cpp [hunk 14]
```diff
@@ -33,34 +33,21 @@ ExpressionContext and_filter_expression_contexts(
...
-                !res.expression_nodes_.contains(intermediate_name),
-                "Expression nodes already contains {}",
-                intermediate_name
-        );
-        res.add_expression_node(intermediate_name, std::make_shared<ExpressionNode>(and_node));
-        overall_root_name = ExpressionName{intermediate_name};
     }
 
-    res.root_node_name_ = *overall_root_name;
+    ExpressionContext res;
```
Comment (alexowens90) [inline cpp/arcticdb/processing/query_planner.cpp:54]: If we keep the `dynamic_schema_` field, then presumably should be inherited from the input expression contexts? Or are they not set yet?

File: python/arcticdb/version_store/processing.py
Comment (alexowens90) [inline python/arcticdb/version_store/processing.py]: Think this is just used by the (Python) `ExpressionNode` class now, so could be a static method

File: python/tests/unit/arcticdb/version_store/test_filtering.py [hunk 15]
```diff
@@ -555,6 +555,64 @@ def test_filter_isin_clashing_sets(
...
+    lib.write(sym, df)
+
+    limit = 3
+    q = QueryBuilder()
+    spread = q["bid"] - q["ask"]
+    q = q[(spread > 0) & (spread < limit)]
+
+    pandas_spread = df["bid"] - df["ask"]
+    expected = df[(pandas_spread > 0) & (pandas_spread < limit)].reset_index(drop=True)
+    received = lib.read(sym, query_builder=q).data
```
Comment (alexowens90) [inline python/tests/unit/arcticdb/version_store/test_filtering.py:612]: This will always be empty as the spread is always `-5`, might be more interesting to test a case with a non-empty result

File: cpp/arcticdb/processing/test/ast_test_helpers.hpp
Comment (alexowens90) [inline cpp/arcticdb/processing/test/ast_test_helpers.hpp:1]: nit: should be in `unit_test_srcs`

File: cpp/arcticdb/processing/expression_node.cpp (hunk 9 above)
Comment (alexowens90) [reply cpp/arcticdb/processing/expression_node.cpp:94]: Ah I see this isn't part of the Python repr, just for the memoization, ignore me

File: cpp/arcticdb/processing/test/test_expression.cpp [hunk 16]
```diff
@@ -28,19 +32,100 @@ TEST(ExpressionNode, AddBasic) {
...
+    for (auto j = 0; j < 20; ++j) {
+        auto v1 = proc.segments_->at(0)->scalar_at<uint64_t>(j, 1).value();
+        auto v2 = proc.segments_->at(0)->scalar_at<uint64_t>(j, 2).value();
+        ASSERT_EQ(result_col->scalar_at<uint64_t>(j), (v1 + v2) * (v1 + v2));
+    }
+
+    const std::string add_label = R"((Column["thing1"] ADD Column["thing2"]))";
+    ASSERT_EQ(add1->label_, add_label);
+    ASSERT_EQ(add2->label_, add_label);
+    ASSERT_EQ(proc.computed_data_.count(add_label), 1);
```
Comment (alexowens90) [inline cpp/arcticdb/processing/test/test_expression.cpp]: nit: `contains` over `count == 1` everywhere

File: cpp/arcticdb/processing/test/test_filter_and_project_sparse.cpp [hunk 17]
```diff
@@ -26,86 +27,47 @@ class FilterProjectSparse : public testing::Test {
...
             std::string_view left_column_name,
             const std::variant<std::string_view, double, std::unordered_set<double>>& right_input, OperationType op
     ) {
-        const std::string root_node_name("binary filter");
-        const std::string value_name("value");
-        const std::string value_set_name("value set");
-        expression_context->root_node_name_ = ExpressionName(root_node_name);
-        std::shared_ptr<ExpressionNode> expression_node;
+        ExpressionChild right;
         util::variant_match(
```
Comment (alexowens90) [inline cpp/arcticdb/processing/test/test_filter_and_project_sparse.cpp]:
```suggestion
        ExpressionChild right = util::variant_match(
```

File: cpp/arcticdb/processing/test/test_expression.cpp [hunk 18]
```diff
@@ -28,19 +32,100 @@ TEST(ExpressionNode, AddBasic) {
...
+    StreamId symbol("test_longhand");
+    auto wrapper =
+            SinkWrapper(symbol, {scalar_field(DataType::UINT64, "thing1"), scalar_field(DataType::UINT64, "thing2")});
+
+    for (auto j = 0; j < 20; ++j) {
+        wrapper.aggregator_.start_row(timestamp(j))([&](auto&& rb) {
+            rb.set_scalar(1, j);
+            rb.set_scalar(2, j + 1);
+        });
+    }
```
Comment (vasil-pashov) [inline cpp/arcticdb/processing/test/test_expression.cpp]: What do you think about using `create_dense_segment` from `segment_generation_utils.hpp` instead? I find it a bit more readable. It'll look something like

```cpp
SegmentInMemory seg = create_dense_segment(
  {scalar_field(DataType::UINT64, "thing1"), scalar_field(DataType::UINT64, "thing2")},
  std::ranges::views::iota<uint64_t, uitnt64_t>(0, 20),
  std::ranges::views::iota<uint64_t, uitnt64_t>(1, 21)
);
```

there's also `slice_data_into_segments` if you want to simulate how arcticdb slices segments (including getting the index as first column in every slice)

File: cpp/arcticdb/processing/expression_node.cpp [hunk 19]
```diff
@@ -75,43 +76,186 @@ ColumnWithStrings::ColumnWithStrings(
...
+    if (std::holds_alternative<ExpressionNode::Leaf>(a.kind_)) {
+        const auto& leaf_a = std::get<ExpressionNode::Leaf>(a.kind_);
+        const auto& leaf_b = std::get<ExpressionNode::Leaf>(b.kind_);
+        if (leaf_a.index() != leaf_b.index()) {
+            return false;
+        }
+        // Now we know that leaf_a and leaf_b hold the same variant type member
+        return util::variant_match(
+                leaf_a,
+                [&leaf_b](const ColumnName& a_col) { return a_col.value == std::get<ColumnName>(leaf_b).value; },
```
Comment (vasil-pashov) [inline cpp/arcticdb/processing/expression_node.cpp:154]: No need to take any action, but this might break if we have columns with the same name. There are already quite a lot of things that don't work properly in this case, so I don't think we should worry about it.

**Review by vasil-pashov — APPROVED** [review]

File: cpp/arcticdb/processing/expression_node.cpp (hunk 10 above)
Comment (poodlewars) [reply cpp/arcticdb/processing/expression_node.cpp:100]: Yeah, everything is built bottom up.

File: cpp/arcticdb/processing/query_planner.cpp (hunk 13 above)
Comment (poodlewars) [reply cpp/arcticdb/processing/query_planner.cpp:40]: FilterClause already has this assertion,

```
    explicit FilterClause(
            std::unordered_set<std::string> input_columns, ExpressionContext expression_context,
            std::optional<PipelineOptimisation> optimisation
    ) :
        expression_context_(std::make_shared<ExpressionContext>(std::move(expression_context))),
        optimisation_(optimisation.value_or(PipelineOptimisation::SPEED)) {
        user_input::check<ErrorCode::E_INVALID_USER_ARGUMENT>(
                expression_context_->root_ && expression_context_->root_->is_operation(),
                "FilterClause AST would produce a column, not a bitset"
        );
        clause_info_.input_columns_ = std::move(input_columns);
    }
```

It works because the bare bool filter gets wrapped in an IDENTITY operation in the Python layer.

Comment (alexowens90) [reply cpp/arcticdb/processing/query_planner.cpp:40]: Ah cool

**Review by alexowens90 — APPROVED** [review]
